kube-runtime/src/observer.rs

```diff
@@ -0,0 +1,114 @@
+use backoff::ExponentialBackoff;
+use futures::{future::ready, Stream, StreamExt};
+use kube_client::{
+    api::{Api, ListParams},
+    core::Resource,
+};
+use serde::de::DeserializeOwned;
+use std::fmt::Debug;
+
+use crate::{
+    utils,
+    watcher::{watcher, Error, Event, Result},
+};
+
+
+/// A simple observer around a watcher with error handling and retry backoff
+///
+/// # Error handling
+///
+/// An `Observer` sets a sensible default backoff policy for all watch events and will retry
+/// (with expotential backoff) from transient `watcher` failures until the retry policy is breached.
+///
+/// To configure the retry policy use `Observer::backoff`.
+///
+/// Note that while all errors will be retried, some represent the need for user action to recover:
+///
+/// - 404 `ErrorResponse`(watching invalid / missing api kind/group for `K`)
+/// - 403 `ErrorResponse` (missing list + watch rbac verbs for `K`)
+pub struct Observer<K>
+where
+    K: Clone + Resource + Send + Sync + 'static,
+{
+    // temporary builder params
+    api: Api<K>,
+    listparams: Option<ListParams>,
+    backoff: Option<ExponentialBackoff>,
+}
+
+impl<K> From<Api<K>> for Observer<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+{
+    fn from(api: Api<K>) -> Observer<K> {
+        Observer::new(api)
+    }
+}
+
+impl<K> Observer<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+{
+    /// Create a Observer on a reflector on a type `K`
+    ///
+    /// Takes an [`Api`] object that determines how the `Observer` listens for changes to the `K`.
+    ///
+    /// The [`ListParams`] controls to the possible subset of objects of `K` that you want to cache.
+    /// For the full set of objects `K` in the given `Api` scope, you can use [`ListParams::default`].
+    #[must_use]
+    pub fn new(api: Api<K>) -> Self {
+        Self {
+            api,
+            listparams: None,
+            backoff: None,
+        }
+    }
+
+    // start the watcher and filter out backoff errors from the stream for a while
+    fn start(self) -> impl Stream<Item = Result<Event<K>>> {
+        let backoff = self.backoff.unwrap_or_else(|| backoff::ExponentialBackoff {
```

Comment:
Need to cap the backoff interval here. Probably a couple of seconds at most.

Reply:
Not sure I agree with the magnitude here. If there's serious api server (hello rancher) issues going on, i'd rather let this follow normal rules and wait several minutes if necessary. Maybe 5 minutes is a good max cap though, that's a common reconciliation time?

Reply:
Should probably look into how the Go clients handle this, but it seems like your API server would have to be pretty overloaded for minutes to start being required.

Reply:
from a quick scan:

- client-go does have a rudimentary [retry pkg](https://pkg.go.dev/k8s.io/client-go/util/retry) analagous to rust's backoff
- can use it [like this](https://github.com/kubernetes-sigs/controller-runtime/blob/11917e9d90625b51e5dab16b4cf20934971e2f08/pkg/envtest/crd.go#L265-L273)
- and a [backoff manager](https://github.com/kubernetes/client-go/blob/10e087ca394e2987f09e759438f9949a746c1ca0/rest/urlbackoff.go#L35)
- reflectors [generally retry immediately](https://github.com/kubernetes/client-go/blob/10e087ca394e2987f09e759438f9949a746c1ca0/tools/cache/reflector.go#L302-L312) - but [backoff some errors](https://github.com/kubernetes/client-go/blob/10e087ca394e2987f09e759438f9949a746c1ca0/tools/cache/reflector.go#L416-L425) with the manager
- shared informers apparently has [some hooks for it as well](https://github.com/kubernetes/client-go/blob/master/tools/cache/shared_informer.go#L169-L182) but couldn't find it in the code
- otherwise [no significant mention of backoff](https://github.com/kubernetes-sigs/controller-runtime/search?q=backoff) (apart from that trivial one) in controller-runtime

Reply:
more scanning, think i found it:
-  [reflectors start a listwatcher with backoff](https://github.com/kubernetes/kubernetes/blob/806e38aeb7bba6625001c8b1edf92425bd9dc62c/staging/src/k8s.io/client-go/tools/cache/reflector.go#L215-L226) and use the generic backoffManager interface to decide how to control it
- [reflector default backoff parameters](https://github.com/kubernetes/client-go/blob/980663e185ab6fc79163b1c2565034f6d58368db/tools/cache/reflector.go#L177-L181)!


interestingly they also support [reading backoff parameters from the environment](https://github.com/kubernetes/client-go/blob/f6ce18ae578c8cca64d14ab9687824d9e1305a67/rest/client.go#L138-L154) for global client defaults, but that's like 5y old so probably an older pattern

Reply:
Also looks like they don't support a give-up-cap at all: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Backoff

Reply:
so based on the [signature of NewExponentialBackoffManager](https://github.com/kubernetes/apimachinery/blob/7149480564b7d40046f19040efa20171b49059ae/pkg/util/wait/wait.go#L328-L331) with ( 800*time.Millisecond, 30*time.Second, 2*time.Minute, 2.0, 1.0, realClock) implies:

- initial backoff: 800ms,
- **max backoff: 30s**
- resetDuration: 2 minutes
- backoffFactor: 2.0
- jitter: 1.0

Reply:
`resetDuration` is interesting.  `backoff` doesn't currently support that at all (and thus, neither do we).

Reply:
Have pinned your branch of backoff and configured `Observer` to default use `ResetTimerBackoff` wrapping `ExtonentialBackoff` with what i think is closest to client-go's reflector params.